auto-instrumentation/httpurlconnection/library/src/main/java/io/opentelemetry/instrumentation/library/httpurlconnection/HttpUrlReplacements.java

```diff
@@ -0,0 +1,421 @@
+            throws IOException {
+        addTraceContextBeforeFirstConnection(c);
+
+        InputStream inputStream;
+        try {
+            inputStream = c.getInputStream();
+        } catch (IOException e) {
+            // TODO: Report this request with a http span;
+            markReported(c);
+            throw e;
+        }
+
+        // TODO: Report this request with a http span;
+        markReported(c);
+
+        return inputStream;
+    }
+
+    public static synchronized InputStream replacementForErrorStream(HttpURLConnection c) {
+        addTraceContextBeforeFirstConnection(c);
+
+        InputStream errorStream = c.getErrorStream();
+
+        // TODO: Report this request with a http span;
+        markReported(c);
+
+        return errorStream;
+    }
+
+    private static synchronized void addTraceContextBeforeFirstConnection(URLConnection c) {
+        if (activeURLConnections.get(c) == null) {
+            activeURLConnections.put(c, new HttpURLConnectionInfo());
+            try {
+                // TODO: Start span and add trace context using
+                // c.setRequestProperty("propertyKey","propertyValue");
+            } catch (Exception e) {
+                // If connection was already made prior to setting this request property,
+                // (which should not happen as we've instrumented all methods that connect)
+                // above call would throw IllegalStateException. Capture and report this
+                // as we need to know if tracing was not set for any request.
+                // TODO: Report this request with a http span;
+                markReported(c);
+            }
+        }
+    }
+
+    private static synchronized void updateLastSeenTime(URLConnection c) {
+        final HttpURLConnectionInfo info = activeURLConnections.get(c);
+        if (info != null) {
+            info.lastSeenTime = System.currentTimeMillis();
+        }
+    }
+
+    private static synchronized void markHarvestable(URLConnection c) {
+        final HttpURLConnectionInfo info = activeURLConnections.get(c);
+        if (info != null) {
+            info.harvestable = true;
+        }
+    }
+
+    private static synchronized void markReported(URLConnection c) {
+        final HttpURLConnectionInfo info = activeURLConnections.get(c);
+        if (info != null) {
+            info.reported = true;
+        }
+    }
+
+    // TODO: Move scheduleConnectionKiller in a different class as it's the only client callable API
+    // in this class
+    public Runnable scheduleConnectionKiller() {
+        return new Runnable() {
+            @Override
+            public void run() {
+                endOngoingConnectionsOlderThan(CONNECTION_KEEP_ALIVE_INTERVAL);
+            }
+
+            @Override
+            public String toString() {
+                return "EndOngoingConnectionsRunnable";
+            }
+        };
+    }
```

Comment:
Is it okay for it to be optional and client's who think they need this can configure it (they will need manual code change to call our APIs)? If not, I can look for the main application activity and use byte buddy to do BCI to schedule this always. Is there any other better ways, to schedule thread runnable using existing open-telemetry codebases?

Reply:
I'm not sure I fully understand what is it that you suggest to make configurable, but generally speaking we can make anything configurable statically, as we did for the [OkHttp instrumentation](https://github.com/open-telemetry/opentelemetry-android/blob/main/auto-instrumentation/okhttp/okhttp-3.0/library/src/main/java/io/opentelemetry/instrumentation/library/okhttp/v3_0/OkHttpInstrumentationConfig.java) for example. Users would have to manually call those functions, setting their config params, and then we'll need to read those values later on.

Reply:
Actually, I wanted to discuss the approach we should take here - Manual vs Automatic scheduling of this thread (which reports connections that are not already reported and idle for given time threshold). 

Ideally, we should proactively **automatically** schedule this thread always to cover all possible scenario and to ensure connections are reported in the end always. 

Without the thread, right now, connections are reported if any exceptions occur and for getInputStream
/getErrorStream method calls assuming this would be called in the end. ( This can be extended to also report connections when getResponseCode method is called). This should mostly cover all scenarios but we can't 100% ensure that these methods are always certainly called. So, thread becomes useful. 

The[ java HttpURLConnection instrumentation](https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/11cac29ba7c0e9e6eeb2a882912a684e3d348fb0/instrumentation/http-url-connection/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/httpurlconnection/HttpUrlConnectionInstrumentation.java#L140) as well only reports when exceptions occur or getInputStream is called, which I think doesn't cover all scenarios. 

I did not find a clean way to define this auto-instrumentation for thread scheduling. Android Apps don't have a method like java main method which would be called always in the starting. We can perhaps target the "onCreate" methods of starting activity or fragment or service but that might not cover all scenarios. Any suggestions for handling this?

Reply:
We've decided to have this as a manual configuration for now and we can improve on it in future.